File: packages/server/src/features/ingest/validator.ts
import { stat } from 'node:fs/promises';
import { extname } from 'node:path';
import { readTags } from '../tags/index.js';
import type { SongTags } from '@sonarly/shared';
import type { AudioMetadata } from '../tags/index.js';

const SUPPORTED_EXTS = new Set(['.mp3', '.flac', '.ogg', '.m4a']);

export interface ValidationResult {
  valid: boolean;
  reason?: 'unsupported_format' | 'unreadable' | 'missing_required_tags';
  tags?: SongTags;
  meta?: AudioMetadata;
  duration?: number;
}

export async function validateIngestFile(filePath: string): Promise<ValidationResult> {
  if (!SUPPORTED_EXTS.has(extname(filePath).toLowerCase())) {
    return { valid: false, reason: 'unsupported_format' };
  }
  try {
    await stat(filePath);
    const meta = await readTags(filePath);
    const tags = meta.tags;
    if (!tags.title || !tags.artist || !tags.album) {
      return { valid: false, reason: 'missing_required_tags', tags, meta };
    }
    return { valid: true, tags, meta, duration: meta.duration };
  } catch {
    return { valid: false, reason: 'unreadable' };
  }
}

File: server/internal/modules/ingest/validator.go
package ingest

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/miquelrosell99/sonarly/server/internal/audio"
	"github.com/miquelrosell99/sonarly/server/internal/modules/library"
)

// Reason is the typed validation failure, mirroring the old ValidationResult
// reason literals (the review UI keys on these strings).
type Reason string

const (
	ReasonUnsupportedFormat   Reason = "unsupported_format"
	ReasonUnreadable          Reason = "unreadable"
	ReasonMissingRequiredTags Reason = "missing_required_tags"
)

// Validation is the per-file verdict the old validateIngestFile returned.
type Validation struct {
	Valid  bool
	Reason Reason
	Meta   *audio.Metadata
}

// ValidateFile ports the old validateIngestFile: extension allowlist, stat,
// audio.ReadMetadata, then the retired server required-tag rule (title + artist +
// album). Title falls back to the filename stem inside ReadMetadata (wire
// parity), so a tagless file fails on the missing artist/album, exactly as
// the old rule played out.
func ValidateFile(filePath string) Validation {
	if !library.AUDIO_EXTS[strings.ToLower(filepath.Ext(filePath))] {
		return Validation{Valid: false, Reason: ReasonUnsupportedFormat}
	}
	if _, err := os.Stat(filePath); err != nil {
		return Validation{Valid: false, Reason: ReasonUnreadable}
	}
	meta, err := audio.ReadMetadata(filePath)
	if err != nil {
		return Validation{Valid: false, Reason: ReasonUnreadable}
	}
	if meta.Title == "" || displayArtist(meta) == "" || meta.Album == "" {
		return Validation{Valid: false, Reason: ReasonMissingRequiredTags, Meta: meta}
	}
	return Validation{Valid: true, Meta: meta}
}

// displayArtist mirrors the retired server tags.artist check: the raw display artist,
// with the split multi-value list as the fallback source.
func displayArtist(meta *audio.Metadata) string {
	if meta.Artist != "" {
		return meta.Artist
	}
	for _, name := range meta.Artists {
		if strings.TrimSpace(name) != "" {
			return name
		}
	}
	return ""
}
